lib/portlayer/storage/vsphere/datastore.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return nil, err
+	}
+
+	task, err := b.SearchDatastore(ctx, path.Join(d.rooturl, p), &spec)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res := info.Result.(types.HostDatastoreBrowserSearchResults)
+	return &res, nil
+}
+
+// LsDirs returns a list of dirents at the given path (relative to root)
+func (d *datastore) LsDirs(ctx context.Context, p string) (*types.ArrayOfHostDatastoreBrowserSearchResults, error) {
+	spec := types.HostDatastoreBrowserSearchSpec{
+		MatchPattern: []string{"*"},
+	}
+
+	b, err := d.ds.Browser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	task, err := b.SearchDatastoreSubFolders(ctx, path.Join(d.rooturl, p), &spec)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res := info.Result.(types.ArrayOfHostDatastoreBrowserSearchResults)
+	return &res, nil
+}
+
+func (d *datastore) Upload(ctx context.Context, r io.Reader, pth string) error {
+	return d.ds.Upload(ctx, r, path.Join(d.rootdir, pth), &soap.DefaultUpload)
+}
+
+func (d *datastore) Download(ctx context.Context, pth string) (io.ReadCloser, error) {
+	rc, _, err := d.ds.Download(ctx, path.Join(d.rootdir, pth), &soap.DefaultDownload)
+	return rc, err
+}
+
+func (d *datastore) Stat(ctx context.Context, pth string) (types.BaseFileInfo, error) {
+	return d.ds.Stat(ctx, path.Join(d.rootdir, pth))
+}
+
+func (d *datastore) Mv(ctx context.Context, fromPath, toPath string) error {
+	from := path.Join(d.rooturl, fromPath)
+	to := path.Join(d.rooturl, toPath)
+	err := tasks.Wait(ctx, func(context.Context) (tasks.Waiter, error) {
+		return d.fm.MoveDatastoreFile(ctx, from, d.s.Datacenter, to, d.s.Datacenter, true)
+	})
+
+	return err
+}
+
+func (d *datastore) IsVSAN(ctx context.Context) bool {
+	dsType, _ := d.ds.Type(ctx)
+	return dsType == types.HostFileSystemVolumeFileSystemTypeVsan
+}
+
+// This creates the root directory in the datastore and sets the rooturl and
+// rootdir in the datastore struct so we can reuse it for other routines.  This
+// handles vsan + vc, vsan + esx, and esx.  The URI conventions are not the
+// same for each and this tries to create the directory and stash the relevant
+// result so the URI doesn't need to be recomputed for every datastore
+// operation.
+func (d *datastore) mkRootDir(ctx context.Context, rootdir string) error {
+
+	// Handle vsan
+	// Vsan will complain if the root dir exists.  Just call it directly and
+	// swallow the error if it's already there.
+	if d.IsVSAN(ctx) {
+		nm := object.NewDatastoreNamespaceManager(d.s.Vim25())
+
+		// This returns the vmfs path (including the datastore and directory
+		// UUIDs).  Use the directory UUID in future operations because it is
+		// the stable path which we can use regardless of vsan state.
+		uuid, err := nm.CreateDirectory(ctx, d.ds, rootdir, "")
+
+		if err != nil {
```

Comment:
I missed this.  This should check if it's a `SoapFault` before descending.  Will fix.